ARROW-15571: [C++] Add min/max binary scalar kernels to execution engine

rtpsw opened the pull request:
This is a proposal for resolving [ARROW-15571](https://issues.apache.org/jira/browse/ARROW-15571). Tests will be added when the resolution approach is agreed.

rok commented:
Thanks for the PR @rtpsw ! I think the min/max features are already covered but sqrt might be interesting.
@cyb70289 would we want a dedicated sqrt kernel?

cyb70289 commented:
> @cyb70289 would we want a dedicated sqrt kernel?

Given the big performance gap of `pow(x, 0.5)` and `sqrt(x)` for floating point inputs, looks a dedicated sqrt kernel is useful.
https://quick-bench.com/q/17n5Kx7lgcNtSA0p6uqSp3ryBu0

rtpsw commented:
> I think the min/max features are already covered

This is being discussed at [ARROW-15571](https://issues.apache.org/jira/browse/ARROW-15571). I reported there on benchmark results showing better performance for the minimum/maximum operations in this PR.

rtpsw commented:
I kept min/max per the discussion in [ARROW-15571](https://issues.apache.org/jira/browse/ARROW-15571). Which units tests are desired? Presumably they should be similar to existing ones - please point to test files I should look into.

cyb70289 commented:
Thanks @rtpsw !

I suggest we first create a solid benchmark for `max_element_wise` before optimizing it. Adhoc benchmark is hard to get right.
E.g., for the benchmark code in [jira comment](https://issues.apache.org/jira/browse/ARROW-15571?focusedCommentId=17487559&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17487559), on my test machine, the speed ratio is 1.4 to 3.7. But if I swap the order (call `maximum` before `max_element_wise`, t0 is still `max_element_wise`, t1 is stil `maximum`), the speed ratio becomes 0.3 to 0.8 (slower). After some refinement it gives about 20% improvement, which looks reasonable.

Another reason is it duplicates `max_element_wise`, but with much fewer data types supported. I think it's better to augment existing kernel than inventing a new one.
cc @lidavidm for comments.

lidavidm commented:
Right, I think here: https://github.com/apache/arrow/blob/26d6e6217ff79451a3fe366bcc88293c7ae67417/cpp/src/arrow/compute/kernels/scalar_compare.cc#L361

We can basically check if `batch.num_values() == 2` and if so, delegate to ScalarBinaryEqualTypes.

rtpsw commented:
> Right, I think here:
> 
> https://github.com/apache/arrow/blob/26d6e6217ff79451a3fe366bcc88293c7ae67417/cpp/src/arrow/compute/kernels/scalar_compare.cc#L361
> 
> We can basically check if `batch.num_values() == 2` and if so, delegate to ScalarBinaryEqualTypes.

I'm not sure what code you meant. I tried the following:
```
    if (batch.num_values() == 2) {
      return ExecBinaryMinMaxScalar::Exec(ctx, options, batch, out);
    }
```
Unfortunately, this led to a large number of complaints form the compiler, e.g., for `struct Minimum`.

Then, I tried the following:
```
  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
    const ElementWiseAggregateOptions& options = MinMaxState::Get(ctx);
    if (batch.num_values() == 2) {
      return ExecBinaryMinMaxScalar<Op>(ctx, options, batch, out);
    }
```
This compiles but, unfortunately, then `arrow-compute-scalar-test` fails with a segmentation fault at TestVarArgsCompareNumeric/0, where TypeParam = arrow::UInt8Type".

I don't know enough to predict the impact of such a change even if made to work. I tend to say it is out of scope for this issue.

lidavidm commented:
Hmm, I'll give it a try when I get a chance - ideally, it should be equivalent to having a specialized kernel, but without having to choose different functions based on input arity

cyb70289 commented:
Benchmarked with PR https://github.com/apache/arrow/pull/12478, this specialized binary min/max improves performance significantly (test set = 32k integers).

When there's no nulls, perf improves from 619 to 1084. Improvement is much bigger with nulls, as binary arithmetic kernel always runs the exec function regardless of nulls to leverage auto vectorization.

max_element_wise(variadic)
```
MaxElementWiseArrayArrayInt64/32768/10000        139 us          139 us         5034 items_per_second=235.612M/s null_percent=0.01 size=32.768k
MaxElementWiseArrayArrayInt64/32768/100          185 us          185 us         3787 items_per_second=177.552M/s null_percent=1 size=32.768k
MaxElementWiseArrayArrayInt64/32768/10           269 us          269 us         2603 items_per_second=121.876M/s null_percent=10 size=32.768k
MaxElementWiseArrayArrayInt64/32768/2            472 us          472 us         1483 items_per_second=69.4022M/s null_percent=50 size=32.768k
MaxElementWiseArrayArrayInt64/32768/1           19.8 us         19.8 us        35449 items_per_second=1.65711G/s null_percent=100 size=32.768k
MaxElementWiseArrayArrayInt64/32768/0           52.9 us         52.9 us        13172 items_per_second=619.058M/s null_percent=0 size=32.768k
```

maximum(binary)
```
MaxElementWiseArrayArrayInt64/32768/10000       30.3 us         30.3 us        23388 items_per_second=1079.83M/s null_percent=0.01 size=32.768k
MaxElementWiseArrayArrayInt64/32768/100         30.3 us         30.3 us        23055 items_per_second=1082.26M/s null_percent=1 size=32.768k
MaxElementWiseArrayArrayInt64/32768/10          30.2 us         30.2 us        23153 items_per_second=1085.95M/s null_percent=10 size=32.768k
MaxElementWiseArrayArrayInt64/32768/2           30.4 us         30.4 us        23105 items_per_second=1079.41M/s null_percent=50 size=32.768k
MaxElementWiseArrayArrayInt64/32768/1           30.1 us         30.1 us        23183 items_per_second=1086.96M/s null_percent=100 size=32.768k
MaxElementWiseArrayArrayInt64/32768/0           30.2 us         30.2 us        23091 items_per_second=1084.69M/s null_percent=0 size=32.768k
```

lidavidm commented:
@cyb70289 thanks for checking this! It should be possible to specialize the variadic kernel for this case (I haven't gotten a chance to look at the details yet though)

rtpsw commented:
Thanks @cyb70289 !

@lidavidm: What do you think remains to be done? Should I add unit tests like in https://github.com/apache/arrow/pull/12412/ ? Or perhaps the binary scalar kernels should be structured differently?

rtpsw commented:
> Another reason is it duplicates `max_element_wise`, but with much fewer data types supported. I think it's better to augment existing kernel than inventing a new one.

IIUC, your suggestion is to add the binary scalar kernels to the existing `max_element_wise` function. I scanned the Arrow codebase looking for a case where a `Function` was added kernels in more than one place, e.g. in `scalar_compare.cc` as in `max_element_wise` and in `scalar_arithmetic.cc` as in `maximum`, but didn't find such a case. I'm not sure how something like this would be implemented. It looks like in one place the function would need to get created with kernels and added to the registry and in another place get fetched from the registry and added additional kernels. This seems to require an annoying change like a predefined order of initialization across the `scalar_*.cc` files or the adding of a get-or-create method to the registry.

lidavidm commented:
Yes, all the kernels need to be added in the same place. However, we don't need to add new kernels, we can branch within the existing kernel. See this sketch (incomplete):

<details>

```diff
diff --git a/cpp/src/arrow/compute/kernels/scalar_compare.cc b/cpp/src/arrow/compute/kernels/scalar_compare.cc
index fc8527810..7ee887233 100644
--- a/cpp/src/arrow/compute/kernels/scalar_compare.cc
+++ b/cpp/src/arrow/compute/kernels/scalar_compare.cc
@@ -70,24 +70,24 @@ struct GreaterEqual {
 
 struct Minimum {
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_floating_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_floating_value<T> Call(KernelContext*, Arg0 left, Arg1 right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::fmin(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_integer_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_integer_value<T> Call(KernelContext*, Arg0 left, Arg1 right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::min(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_decimal_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_decimal_value<T> Call(KernelContext*, Arg0 left, Arg1 right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::min(left, right);
   }
 
-  static string_view Call(string_view left, string_view right) {
+  static string_view Call(KernelContext*, string_view left, string_view right, Status*) {
     return std::min(left, right);
   }
 
@@ -114,24 +114,24 @@ struct Minimum {
 
 struct Maximum {
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_floating_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_floating_value<T> Call(KernelContext*, Arg0 left, Arg1 right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::fmax(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_integer_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_integer_value<T> Call(KernelContext*, Arg0 left, Arg1 right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::max(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_decimal_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_decimal_value<T> Call(KernelContext*, Arg0 left, Arg1 right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::max(left, right);
   }
 
-  static string_view Call(string_view left, string_view right) {
+  static string_view Call(KernelContext*, string_view left, string_view right, Status*) {
     return std::max(left, right);
   }
 
@@ -330,11 +330,12 @@ template <typename OutType, typename Op>
 struct ScalarMinMax {
   using OutValue = typename GetOutputType<OutType>::T;
 
-  static void ExecScalar(const ExecBatch& batch,
+  static Status ExecScalar(KernelContext* ctx, const ExecBatch& batch,
                          const ElementWiseAggregateOptions& options, Scalar* out) {
     // All arguments are scalar
     OutValue value{};
     bool valid = false;
+    Status st = Status::OK();
     for (const auto& arg : batch.values) {
       // Ignore non-scalar arguments so we can use it in the mixed-scalar-and-array case
       if (!arg.is_scalar()) continue;
@@ -342,31 +343,63 @@ struct ScalarMinMax {
       if (!scalar.is_valid) {
         if (options.skip_nulls) continue;
         out->is_valid = false;
-        return;
+        return st;
       }
       if (!valid) {
         value = UnboxScalar<OutType>::Unbox(scalar);
         valid = true;
       } else {
         value = Op::template Call<OutValue, OutValue, OutValue>(
-            value, UnboxScalar<OutType>::Unbox(scalar));
+            ctx, value, UnboxScalar<OutType>::Unbox(scalar), &st);
+        RETURN_NOT_OK(st);
       }
     }
     out->is_valid = valid;
     if (valid) {
       BoxScalar<OutType>::Box(value, out);
     }
+    return st;
+  }
+
+  // Specialization for 2-ary case with skip_nulls
+  static Status ExecBinary(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (out->is_scalar()) {
+      out->scalar()->is_valid = batch[0].scalar()->is_valid && batch[1].scalar()->is_valid;
+    } else if (batch[0].is_scalar()) {
+      ArrayData* output = out->mutable_array();
+      // TODO: if scalar is invalid, allocate all-null bitmap; else:
+      // if batch[1].MayHaveNulls(), copy its bitmap (batch[1] should
+      // not be a scalar since then we'd be in the first case above)
+    } else if (batch[1].is_scalar()) {
+      // TODO:
+    } else {
+      ArrayData* output = out->mutable_array();
+      const ArrayData& left = *batch[0].array();
+      const ArrayData& right = *batch[1].array();
+      // TODO: note that left and/or right may not have a validity
+      // bitmap
+      ARROW_ASSIGN_OR_RAISE(output->buffers[0], ::arrow::internal::BitmapAnd(
+          ctx->memory_pool(),
+          left.GetValues<uint8_t>(0, 0), left.offset,
+          right.GetValues<uint8_t>(0, 0), right.offset,
+          batch.length, /*out_offset=*/0));
+    }
+    return applicator::ScalarBinaryEqualTypes<OutType, OutType, Op>::Exec(ctx, batch, out);
   }
 
   static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
     const ElementWiseAggregateOptions& options = MinMaxState::Get(ctx);
+
+    if (batch.num_values() == 2 && options.skip_nulls) {
+      return ExecBinary(ctx, batch, out);
+    }
+
     const auto descrs = batch.GetDescriptors();
     const size_t scalar_count =
         static_cast<size_t>(std::count_if(batch.values.begin(), batch.values.end(),
                                           [](const Datum& d) { return d.is_scalar(); }));
     if (scalar_count == batch.values.size()) {
-      ExecScalar(batch, options, out->scalar().get());
-      return Status::OK();
+      return ExecScalar(ctx, batch, options, out->scalar().get());
     }
 
     ArrayData* output = out->mutable_array();
@@ -382,7 +415,7 @@ struct ScalarMinMax {
     if (scalar_count > 0) {
       ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> temp_scalar,
                             MakeScalar(out->type(), 0));
-      ExecScalar(batch, options, temp_scalar.get());
+      RETURN_NOT_OK(ExecScalar(ctx, batch, options, temp_scalar.get()));
       if (temp_scalar->is_valid) {
         const auto value = UnboxScalar<OutType>::Unbox(*temp_scalar);
         initialize_output = false;
@@ -444,6 +477,7 @@ struct ScalarMinMax {
       }
     }
 
+    Status st = Status::OK();
     for (const auto& array : arrays) {
       OutputArrayWriter<OutType> writer(out->mutable_array());
       ArrayIterator<OutType> out_it(*output);
@@ -454,7 +488,7 @@ struct ScalarMinMax {
             auto u = out_it();
             if (!output->buffers[0] ||
                 bit_util::GetBit(output->buffers[0]->data(), index)) {
-              writer.Write(Op::template Call<OutValue, OutValue, OutValue>(u, value));
+              writer.Write(Op::template Call<OutValue, OutValue, OutValue>(ctx, u, value, &st));
             } else {
               writer.Write(value);
             }
@@ -468,7 +502,7 @@ struct ScalarMinMax {
           });
     }
     output->null_count = output->buffers[0] ? -1 : 0;
-    return Status::OK();
+    return st;
   }
 };
 
@@ -492,6 +526,7 @@ Status ExecBinaryMinMaxScalar(KernelContext* ctx,
   const auto& first_scalar = *batch.values.front().scalar();
   string_view result = checked_cast<const BaseBinaryScalar&>(first_scalar).view();
   bool valid = first_scalar.is_valid;
+  Status st = Status::OK();
   for (size_t i = 1; i < batch.values.size(); i++) {
     const auto& scalar = *batch[i].scalar();
     if (!scalar.is_valid) {
@@ -499,7 +534,7 @@ Status ExecBinaryMinMaxScalar(KernelContext* ctx,
       continue;
     } else {
       string_view value = checked_cast<const BaseBinaryScalar&>(scalar).view();
-      result = !valid ? value : Op::Call(result, value);
+      result = !valid ? value : Op::Call(ctx, result, value, &st);
       valid = true;
     }
   }
@@ -510,7 +545,7 @@ Status ExecBinaryMinMaxScalar(KernelContext* ctx,
   } else {
     output->is_valid = false;
   }
-  return Status::OK();
+  return st;
 }
 
 template <typename Type, typename Op>
@@ -537,10 +572,11 @@ struct BinaryScalarMinMax {
     RETURN_NOT_OK(builder.Reserve(batch.length));
     RETURN_NOT_OK(builder.ReserveData(estimated_final_size));
 
+    Status st = Status::OK();
     for (int64_t row = 0; row < batch.length; row++) {
       util::optional<string_view> result;
       auto visit_value = [&](string_view value) {
-        result = !result ? value : Op::Call(*result, value);
+        result = !result ? value : Op::Call(ctx, *result, value, &st);
       };
 
       for (size_t col = 0; col < batch.values.size(); col++) {
@@ -548,6 +584,7 @@ struct BinaryScalarMinMax {
           const auto& scalar = *batch[col].scalar();
           if (scalar.is_valid) {
             visit_value(UnboxScalar<Type>::Unbox(scalar));
+            RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = util::nullopt;
             break;
@@ -561,6 +598,7 @@ struct BinaryScalarMinMax {
             const int64_t length = offsets[row + 1] - offsets[row];
             visit_value(
                 string_view(reinterpret_cast<const char*>(data + offsets[row]), length));
+            RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = util::nullopt;
             break;
@@ -629,10 +667,11 @@ struct FixedSizeBinaryScalarMinMax {
     RETURN_NOT_OK(builder.ReserveData(estimated_final_size));
 
     std::vector<string_view> valid_cols(batch.values.size());
+    Status st = Status::OK();
     for (int64_t row = 0; row < batch.length; row++) {
       string_view result;
       auto visit_value = [&](string_view value) {
-        result = result.empty() ? value : Op::Call(result, value);
+        result = result.empty() ? value : Op::Call(ctx, result, value, &st);
       };
 
       for (size_t col = 0; col < batch.values.size(); col++) {
@@ -640,6 +679,7 @@ struct FixedSizeBinaryScalarMinMax {
           const auto& scalar = *batch[col].scalar();
           if (scalar.is_valid) {
             visit_value(UnboxScalar<FixedSizeBinaryType>::Unbox(scalar));
+            RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = string_view();
             break;
@@ -651,6 +691,7 @@ struct FixedSizeBinaryScalarMinMax {
             const auto data = array.GetValues<uint8_t>(1, /*absolute_offset=*/0);
             visit_value(string_view(
                 reinterpret_cast<const char*>(data) + row * byte_width, byte_width));
+            RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = string_view();
             break;
@@ -699,12 +740,14 @@ std::shared_ptr<ScalarFunction> MakeScalarMinMax(std::string name,
   auto func = std::make_shared<VarArgsCompareFunction>(
       name, Arity::VarArgs(), doc, &default_element_wise_aggregate_options);
   for (const auto& ty : NumericTypes()) {
-    auto exec = GeneratePhysicalNumeric<ScalarMinMax, Op>(ty);
-    ScalarKernel kernel{KernelSignature::Make({ty}, ty, /*is_varargs=*/true), exec,
-                        MinMaxState::Init};
-    kernel.null_handling = NullHandling::type::COMPUTED_NO_PREALLOCATE;
-    kernel.mem_allocation = MemAllocation::type::PREALLOCATE;
-    DCHECK_OK(func->AddKernel(std::move(kernel)));
+    {
+      auto exec = GeneratePhysicalNumeric<ScalarMinMax, Op>(ty);
+      ScalarKernel kernel{KernelSignature::Make({ty}, ty, /*is_varargs=*/true), exec,
+        MinMaxState::Init};
+      kernel.null_handling = NullHandling::type::COMPUTED_NO_PREALLOCATE;
+      kernel.mem_allocation = MemAllocation::type::PREALLOCATE;
+      DCHECK_OK(func->AddKernel(std::move(kernel)));
+    }
   }
   for (const auto& ty : TemporalTypes()) {
     auto exec = GeneratePhysicalNumeric<ScalarMinMax, Op>(ty);
```

</details>

By the way, I note on the JIRA that we want to return NaN if only one NaN input is provided. Is this not already the case?

```python
>>> pc.min_element_wise(pa.scalar(float('nan')))
<pyarrow.DoubleScalar: nan>
>>> pc.min_element_wise([float('nan')])
<pyarrow.lib.DoubleArray object at 0x7f1693ad7e80>
[
  nan
]
>>> pc.min_element_wise([float('nan')], pa.array([None], 'float32'))
<pyarrow.lib.DoubleArray object at 0x7f1693b40160>
[
  nan
]
>>> pc.min_element_wise(pa.scalar(float('nan')), pa.scalar(None, 'float32'))
<pyarrow.DoubleScalar: nan>
```

rtpsw commented:
OK, I'll take a stab at implementing this branching.

rtpsw commented:
I tried the code included below. It compiles but fails the `TestVarArgsCompareNumeric/0.MinElementWise` test of `arrow-compute-scalar-test`. Because I have less time to spend on this issue right now, I thought I'd post the code in case the errors are obvious to someone.

<details><summary>Details</summary><p>

```
diff --git a/cpp/src/arrow/compute/kernels/scalar_compare.cc b/cpp/src/arrow/compute/kernels/scalar_compare.cc
index fc8527810..30e1db022 100644
--- a/cpp/src/arrow/compute/kernels/scalar_compare.cc
+++ b/cpp/src/arrow/compute/kernels/scalar_compare.cc
@@ -70,24 +70,24 @@ struct GreaterEqual {
 
 struct Minimum {
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_floating_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_floating_value<T> Call(KernelContext*, const Arg0& left, const Arg1& right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::fmin(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_integer_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_integer_value<T> Call(KernelContext*, const Arg0& left, const Arg1& right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::min(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_decimal_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_decimal_value<T> Call(KernelContext*, const Arg0& left, const Arg1& right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::min(left, right);
   }
 
-  static string_view Call(string_view left, string_view right) {
+  static string_view Call(KernelContext*, const string_view& left, const string_view& right, Status*) {
     return std::min(left, right);
   }
 
@@ -114,24 +114,24 @@ struct Minimum {
 
 struct Maximum {
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_floating_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_floating_value<T> Call(KernelContext*, const Arg0& left, const Arg1& right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::fmax(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_integer_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_integer_value<T> Call(KernelContext*, const Arg0& left, const Arg1& right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::max(left, right);
   }
 
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_decimal_value<T> Call(Arg0 left, Arg1 right) {
+  static enable_if_decimal_value<T> Call(KernelContext*, const Arg0& left, const Arg1& right, Status*) {
     static_assert(std::is_same<T, Arg0>::value && std::is_same<Arg0, Arg1>::value, "");
     return std::max(left, right);
   }
 
-  static string_view Call(string_view left, string_view right) {
+  static string_view Call(KernelContext*, const string_view& left, const string_view& right, Status*) {
     return std::max(left, right);
   }
 
@@ -330,11 +330,12 @@ template <typename OutType, typename Op>
 struct ScalarMinMax {
   using OutValue = typename GetOutputType<OutType>::T;
 
-  static void ExecScalar(const ExecBatch& batch,
-                         const ElementWiseAggregateOptions& options, Scalar* out) {
+  static Status ExecScalar(KernelContext* ctx, const ExecBatch& batch,
+                           const ElementWiseAggregateOptions& options, Scalar* out) {
     // All arguments are scalar
     OutValue value{};
     bool valid = false;
+    Status st = Status::OK();
     for (const auto& arg : batch.values) {
       // Ignore non-scalar arguments so we can use it in the mixed-scalar-and-array case
       if (!arg.is_scalar()) continue;
@@ -342,31 +343,97 @@ struct ScalarMinMax {
       if (!scalar.is_valid) {
         if (options.skip_nulls) continue;
         out->is_valid = false;
-        return;
+        return st;
       }
       if (!valid) {
         value = UnboxScalar<OutType>::Unbox(scalar);
         valid = true;
       } else {
         value = Op::template Call<OutValue, OutValue, OutValue>(
-            value, UnboxScalar<OutType>::Unbox(scalar));
+            ctx, value, UnboxScalar<OutType>::Unbox(scalar), &st);
+       RETURN_NOT_OK(st);
       }
     }
     out->is_valid = valid;
     if (valid) {
       BoxScalar<OutType>::Box(value, out);
     }
+    return st;
+  }
+
+  // Specialization for 2-ary case with skip_nulls
+  static Status ExecBinary(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch.num_values() != 2) {
+      return Status::Invalid("Number of batch values must be 2 but is ",
+                             batch.num_values());
+    }
+    if (out->is_scalar()) {
+      out->scalar()->is_valid =
+          batch[0].scalar()->is_valid && batch[1].scalar()->is_valid;
+    } else if (batch[0].is_scalar() || batch[1].is_scalar()) {
+      const ArrayData& array = *(batch[0].is_scalar() ? batch[1] : batch[0]).array();
+      ArrayData* output = out->mutable_array();
+      // TODO: if scalar is invalid, allocate all-null bitmap; else:
+      // if batch[1].MayHaveNulls(), copy its bitmap (batch[1] should
+      // not be a scalar since then we'd be in the first case above)
+      if (!batch[0].scalar()->is_valid) {
+        //output->buffers[0] = NULLPTR;
+        ARROW_ASSIGN_OR_RAISE(
+            output->buffers[0], ::arrow::AllocateEmptyBitmap(
+                batch.length, ctx->memory_pool()));
+      } else if (array.MayHaveNulls()) {
+        ARROW_ASSIGN_OR_RAISE(
+            output->buffers[0], ::arrow::Buffer::Copy(
+                array.buffers[0], array.buffers[0]->memory_manager()));
+      } else {
+        ARROW_ASSIGN_OR_RAISE(
+            output->buffers[0], ::arrow::AllocateEmptyBitmap(
+                batch.length, ctx->memory_pool()));
+       ::arrow::internal::InvertBitmap(
+            output->buffers[0]->mutable_data(), 0,
+            batch.length,
+            output->buffers[0]->mutable_data(), 0);
+      }
+    } else {
+      ArrayData* output = out->mutable_array();
+      const ArrayData& left = *batch[0].array();
+      const ArrayData& right = *batch[1].array();
+      if (left.MayHaveNulls() && right.MayHaveNulls()) {
+        // TODO: note that left and/or right may not have a validity
+        // bitmap
+        ARROW_ASSIGN_OR_RAISE(output->buffers[0], ::arrow::internal::BitmapAnd(
+            ctx->memory_pool(),
+            left.GetValues<uint8_t>(0, 0), left.offset,
+            right.GetValues<uint8_t>(0, 0), right.offset,
+            batch.length, /*out_offset=*/0));
+      } else if (!left.MayHaveNulls() && !right.MayHaveNulls()) {
+        //output->buffers[0] = NULLPTR;
+        ARROW_ASSIGN_OR_RAISE(
+            output->buffers[0], ::arrow::AllocateEmptyBitmap(
+                batch.length, ctx->memory_pool()));
+      } else {
+        const ArrayData& array = !left.MayHaveNulls() ? right : left;
+        ARROW_ASSIGN_OR_RAISE(
+            output->buffers[0], ::arrow::Buffer::Copy(
+                array.buffers[0], array.buffers[0]->memory_manager()));
+      }
+    }
+    return applicator::ScalarBinaryEqualTypes<OutType, OutType, Op>::Exec(ctx, batch, out);
   }
 
   static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
     const ElementWiseAggregateOptions& options = MinMaxState::Get(ctx);
+
+    if (batch.num_values() == 2 && options.skip_nulls) {
+      return ExecBinary(ctx, batch, out);
+    }
+
     const auto descrs = batch.GetDescriptors();
     const size_t scalar_count =
         static_cast<size_t>(std::count_if(batch.values.begin(), batch.values.end(),
                                           [](const Datum& d) { return d.is_scalar(); }));
     if (scalar_count == batch.values.size()) {
-      ExecScalar(batch, options, out->scalar().get());
-      return Status::OK();
+      return ExecScalar(ctx, batch, options, out->scalar().get());
     }
 
     ArrayData* output = out->mutable_array();
@@ -382,7 +449,7 @@ struct ScalarMinMax {
     if (scalar_count > 0) {
       ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> temp_scalar,
                             MakeScalar(out->type(), 0));
-      ExecScalar(batch, options, temp_scalar.get());
+      RETURN_NOT_OK(ExecScalar(ctx, batch, options, temp_scalar.get()));
       if (temp_scalar->is_valid) {
         const auto value = UnboxScalar<OutType>::Unbox(*temp_scalar);
         initialize_output = false;
@@ -444,6 +511,7 @@ struct ScalarMinMax {
       }
     }
 
+    Status st = Status::OK();
     for (const auto& array : arrays) {
       OutputArrayWriter<OutType> writer(out->mutable_array());
       ArrayIterator<OutType> out_it(*output);
@@ -454,7 +522,7 @@ struct ScalarMinMax {
             auto u = out_it();
             if (!output->buffers[0] ||
                 bit_util::GetBit(output->buffers[0]->data(), index)) {
-              writer.Write(Op::template Call<OutValue, OutValue, OutValue>(u, value));
+              writer.Write(Op::template Call<OutValue, OutValue, OutValue>(ctx, u, value, &st));
             } else {
               writer.Write(value);
             }
@@ -468,7 +536,7 @@ struct ScalarMinMax {
           });
     }
     output->null_count = output->buffers[0] ? -1 : 0;
-    return Status::OK();
+    return st;
   }
 };
 
@@ -492,6 +560,7 @@ Status ExecBinaryMinMaxScalar(KernelContext* ctx,
   const auto& first_scalar = *batch.values.front().scalar();
   string_view result = checked_cast<const BaseBinaryScalar&>(first_scalar).view();
   bool valid = first_scalar.is_valid;
+  Status st = Status::OK();
   for (size_t i = 1; i < batch.values.size(); i++) {
     const auto& scalar = *batch[i].scalar();
     if (!scalar.is_valid) {
@@ -499,7 +568,7 @@ Status ExecBinaryMinMaxScalar(KernelContext* ctx,
       continue;
     } else {
       string_view value = checked_cast<const BaseBinaryScalar&>(scalar).view();
-      result = !valid ? value : Op::Call(result, value);
+      result = !valid ? value : Op::Call(ctx, result, value, &st);
       valid = true;
     }
   }
@@ -510,7 +579,7 @@ Status ExecBinaryMinMaxScalar(KernelContext* ctx,
   } else {
     output->is_valid = false;
   }
-  return Status::OK();
+  return st;
 }
 
 template <typename Type, typename Op>
@@ -537,10 +606,11 @@ struct BinaryScalarMinMax {
     RETURN_NOT_OK(builder.Reserve(batch.length));
     RETURN_NOT_OK(builder.ReserveData(estimated_final_size));
 
+    Status st = Status::OK();
     for (int64_t row = 0; row < batch.length; row++) {
       util::optional<string_view> result;
       auto visit_value = [&](string_view value) {
-        result = !result ? value : Op::Call(*result, value);
+        result = !result ? value : Op::Call(ctx, *result, value, &st);
       };
 
       for (size_t col = 0; col < batch.values.size(); col++) {
@@ -548,6 +618,7 @@ struct BinaryScalarMinMax {
           const auto& scalar = *batch[col].scalar();
           if (scalar.is_valid) {
             visit_value(UnboxScalar<Type>::Unbox(scalar));
+           RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = util::nullopt;
             break;
@@ -561,6 +632,7 @@ struct BinaryScalarMinMax {
             const int64_t length = offsets[row + 1] - offsets[row];
             visit_value(
                 string_view(reinterpret_cast<const char*>(data + offsets[row]), length));
+           RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = util::nullopt;
             break;
@@ -629,10 +701,11 @@ struct FixedSizeBinaryScalarMinMax {
     RETURN_NOT_OK(builder.ReserveData(estimated_final_size));
 
     std::vector<string_view> valid_cols(batch.values.size());
+    Status st = Status::OK();
     for (int64_t row = 0; row < batch.length; row++) {
       string_view result;
       auto visit_value = [&](string_view value) {
-        result = result.empty() ? value : Op::Call(result, value);
+        result = result.empty() ? value : Op::Call(ctx, result, value, &st);
       };
 
       for (size_t col = 0; col < batch.values.size(); col++) {
@@ -640,6 +713,7 @@ struct FixedSizeBinaryScalarMinMax {
           const auto& scalar = *batch[col].scalar();
           if (scalar.is_valid) {
             visit_value(UnboxScalar<FixedSizeBinaryType>::Unbox(scalar));
+           RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = string_view();
             break;
@@ -651,6 +725,7 @@ struct FixedSizeBinaryScalarMinMax {
             const auto data = array.GetValues<uint8_t>(1, /*absolute_offset=*/0);
             visit_value(string_view(
                 reinterpret_cast<const char*>(data) + row * byte_width, byte_width));
+           RETURN_NOT_OK(st);
           } else if (!options.skip_nulls) {
             result = string_view();
             break;
```
</p>
</details>

lidavidm commented:
I will take a look (though it may take me a while as well).

There's a larger meta-question here: I think we should consider adding specialization to the function registry, to make this easier. A lot of the details here are in just replicating allocation/null bitmap handling policies that the kernels framework already handles for us. But right now, you can't add kernels that don't have the same arity as the function, it would be useful if (for at least variadic kernels) we could directly register 1-adic, 2-adic, etc. specializations.

lidavidm commented:
@pitrou @edponce you may have opinions on the above (allowing variadic functions to register fixed-arity specializations directly), I tried that and this check stops it: https://github.com/apache/arrow/blob/094c5ba186cddd69d4aa83de5ed2b62d4ed07081/cpp/src/arrow/compute/function.cc#L320-L322

pitrou commented:
> But right now, you can't add kernels that don't have the same arity as the function, it would be useful if (for at least variadic kernels) we could directly register 1-adic, 2-adic, etc. specializations.

That sounds interesting, but how would you distinguish a fixed-args kernel from a variable-args kernel?

lidavidm commented:
The infra just iterates through each kernel in turn anyways, so it would take the first kernel that matches. Hence if two parameters got passed, and the 2-adic kernels were listed first, it would match; if three arguments were passed, it wouldn't match the arity of the 2-adic kernel and would fall through to the variadic kernel.

pitrou commented:
Hmm, I see. I don't know if we want to rely on registration order, though?

lidavidm commented:
True, registration order might not be the best way to do it. We could just sort kernels based on arity?

pitrou commented:
Or perhaps have an attribute to mark a kernel var-arg? If you have two 2-argument kernels for a vararg function, but you're calling the function with 3 arguments, you don't want to invoke the kernel that's specialized for 2 arguments.

lidavidm commented:
Oh, kernels already have an InputType so we know whether it's variadic or not already.

lidavidm commented:
So we could just sort fixed-arity kernels before variadic kernels and it would give us the behavior we want, I think.

pitrou commented:
You mean sort them at registration time? That sounds better than at kernel lookup time.

lidavidm commented:
Yes (or really at AddFunction time presumably)

edponce commented:
A related topic is having kernel specializations based on function options or other non-type factors (template specialization is used for type-specific cases). [For the `binary_repeat` function, there are two specializations based on the number of repeats requested. The dispatching is done at runtime](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_string_ascii.cc#L3462-L3467). Could an approach similar to this be applied here?

lidavidm commented:
Arity is a fairly easy one just because we effectively already are doing it here: https://github.com/apache/arrow/blob/9df9bad7d1c37971dab03361df155ce1709fe622/cpp/src/arrow/compute/function.cc#L129-L134

Once we get beyond that it seems more complicated; I wonder if logic like that could be done in DispatchBest somehow though.

edponce commented:
I agree with sorting on a per-kernel basis based on arity (variadic being last case) during kernel registration. I think this is a good solution.

lidavidm commented:
Sorry for the delay. I gave this a try, but @edponce is right: even for this, we need specialization based on function options too, since the 2-adic kernel only works when `options.skip_nulls`.

edponce commented:
We can probably add `FunctionOptions` to the kernel dispatching mechanism (in another PR), and have a kernel-specific lambda/function resolve the kernels based on the options. Need to sketch how something like this would look.

In the meantime, resolving kernel at runtime should be ok.

lidavidm commented:
Sorry, this has mostly slipped off my queue - I will try to get back to you soon. I think in the short term we just need to do the dispatching in the kernel itself. I'm not sure if we have enough functions that need to do dispatching like this to warrant a more complex dispatching mechanism built into the functions framework, it would clean up a few functions but being able to dispatch on options seems potentially complicated

amol- commented:
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍
